transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java

```diff
@@ -111,82 +123,120 @@ private boolean enqueueSqe(int op, int rwFlags, int fd, long bufferAddress, int
+    boolean addPoll(int fd, int pollMask) {
         return enqueueSqe(Native.IORING_OP_POLL_ADD, pollMask, fd, 0, 0, 0);
     }
 
-    //return true -> submit() was called
-    public boolean addRead(int fd, long bufferAddress, int pos, int limit) {
-        return enqueueSqe(Native.IORING_OP_READ, 0, fd, bufferAddress + pos, limit - pos, 0);
+    public void addRead(int fd, long bufferAddress, int pos, int limit, short bufIndex,
+            boolean blocking) {
+        if (bufIndex >= 0) {
+            enqueueFixedSqe(Native.IORING_OP_READ_FIXED, fd,
+                    bufferAddress + pos, limit - pos, bufIndex);
+        } else {
+            enqueueSqe(Native.IORING_OP_READ, 0, fd, bufferAddress + pos, limit - pos, 0);
+        }
+        if (!blocking) {
+            ioInflightCount++;
+        }
     }
 
-    public boolean addWrite(int fd, long bufferAddress, int pos, int limit) {
-        return enqueueSqe(Native.IORING_OP_WRITE, 0, fd, bufferAddress + pos, limit - pos, 0);
+    public void addWrite(int fd, long bufferAddress, int pos, int limit, short bufIndex) {
+        if (bufIndex >= 0) {
+            enqueueFixedSqe(Native.IORING_OP_WRITE_FIXED, fd,
+                    bufferAddress + pos, limit - pos, bufIndex);
+        } else {
+            enqueueSqe(Native.IORING_OP_WRITE, 0, fd, bufferAddress + pos, limit - pos, 0);
+        }
+        ioInflightCount++;
     }
 
-    public boolean addAccept(int fd, long address, long addressLength) {
-        return enqueueSqe(Native.IORING_OP_ACCEPT, Native.SOCK_NONBLOCK | Native.SOCK_CLOEXEC, fd,
+    public void addAccept(int fd, long address, long addressLength) {
+        enqueueSqe(Native.IORING_OP_ACCEPT, Native.SOCK_NONBLOCK | Native.SOCK_CLOEXEC, fd,
                 address, 0, addressLength);
+        ioInflightCount++;
     }
 
     //fill the address which is associated with server poll link user_data
-    public boolean addPollRemove(int fd, int pollMask) {
-        return enqueueSqe(Native.IORING_OP_POLL_REMOVE, 0, fd,
+    public void addPollRemove(int fd, int pollMask) {
+        enqueueSqe(Native.IORING_OP_POLL_REMOVE, 0, fd,
                 convertToUserData(Native.IORING_OP_POLL_ADD, fd, pollMask), 0, 0);
     }
 
-    public boolean addConnect(int fd, long socketAddress, long socketAddressLength) {
-        return enqueueSqe(Native.IORING_OP_CONNECT, 0, fd, socketAddress, 0, socketAddressLength);
+    public void addReadCancel(int fd) {
+        enqueueSqe(Native.IORING_OP_ASYNC_CANCEL, 0, fd,
+                convertToUserData(Native.IORING_OP_READ, fd, 0), 0, 0);
     }
 
-    public boolean addWritev(int fd, long iovecArrayAddress, int length) {
-        return enqueueSqe(Native.IORING_OP_WRITEV, 0, fd, iovecArrayAddress, length, 0);
+    public void addConnect(int fd, long socketAddress, long socketAddressLength) {
+        enqueueSqe(Native.IORING_OP_CONNECT, 0, fd, socketAddress, 0, socketAddressLength);
+        ioInflightCount++;
```

Comment:
why incrementation? as connect,accept..pollRemove event dont use any buffer

Reply:
Though they don't use a buffer, the buffer unregistration/registration syscalls will still block while there are _any_ pending completions (even timeout). This counter is only used to wait until there are no "real" IO ops in progress, at which point it cancels the polls/timeout/blocking eventfd read before making the register syscalls.

Reply:
@njhill that's annoying..maybe that's reason why it is not much faster..I guess
the good thing with registered fd IORING_REGISTER_FILES_UPDATE you able to update without any limitation

Reply:
> maybe that's reason why it is not much faster.

I don't _think_ so because there would be very few of these re-registrations occurring, especially once the app reaches steady-state (the buffer pools grow to appropriate size and then remain mostly static)